File: frontend/src/tests/contract/test_media_generation_post.ts
/**
 * Contract test for POST /api/media/generate endpoint
 *
 * This test validates that the media generation endpoint accepts the new
 * Gemini model selection parameters and returns the expected job response.
 */

import axios from 'axios';
import { MediaGenerateRequest, MediaGenerateResponse } from '../../services/api';

// Mock axios for testing
jest.mock('axios');
const mockedAxios = axios as jest.Mocked<typeof axios>;

describe('Contract: POST /api/media/generate', () => {
  const baseURL = 'http://localhost:8000';
  const endpoint = '/api/media/generate';

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should accept enhanced MediaGenerateRequest with model selection', async () => {
    // Arrange
    const request: MediaGenerateRequest = {
      script_content: 'Speaker 1: Today we explore AI technologies...',
      asset_types: ['image', 'video_clip'],
      num_assets: 5,
      preferred_model: 'gemini-2.5-flash-image',
      allow_fallback: true
    };

    const expectedResponse: MediaGenerateResponse = {
      job_id: '550e8400-e29b-41d4-a716-446655440000',
      status: 'pending',
      model_selected: 'gemini-2.5-flash-image',
      estimated_completion: '2025-09-26T10:35:00Z',
      fallback_occurred: false
    };

    mockedAxios.post.mockResolvedValueOnce({
      status: 202,
      data: expectedResponse
    });

    // Act
    const response = await axios.post(`${baseURL}${endpoint}`, request);

    // Assert
    expect(response.status).toBe(202);
    expect(response.data).toMatchObject(expectedResponse);

    // Validate request structure was sent
    expect(mockedAxios.post).toHaveBeenCalledWith(`${baseURL}${endpoint}`, request);
  });

  it('should validate required request fields', async () => {
    // Arrange - minimal valid request
    const minimalRequest = {
      script_content: 'Test script content',
      asset_types: ['image']
    };

    const response = {
      job_id: '550e8400-e29b-41d4-a716-446655440001',
      status: 'pending',
      model_selected: 'gemini-2.5-flash-image'
    };

    mockedAxios.post.mockResolvedValueOnce({
      status: 202,
      data: response
    });

    // Act
    const result = await axios.post(`${baseURL}${endpoint}`, minimalRequest);

    // Assert
    expect(result.status).toBe(202);
    expect(result.data).toHaveProperty('job_id');
    expect(result.data).toHaveProperty('status');
    expect(result.data).toHaveProperty('model_selected');
  });

  it('should handle fallback model selection', async () => {
    // Arrange - request with fallback enabled
    const requestWithFallback: MediaGenerateRequest = {
      script_content: 'Test content',
      asset_types: ['image'],
      preferred_model: 'gemini-2.5-flash-image',
      allow_fallback: true
    };

    const fallbackResponse: MediaGenerateResponse = {
      job_id: '550e8400-e29b-41d4-a716-446655440002',
      status: 'pending',
      model_selected: 'gemini-pro',
      estimated_completion: '2025-09-26T10:35:00Z',
      fallback_occurred: true
    };

    mockedAxios.post.mockResolvedValueOnce({
      status: 202,
      data: fallbackResponse
    });

    // Act
    const response = await axios.post(`${baseURL}${endpoint}`, requestWithFallback);

    // Assert
    expect(response.data.model_selected).toBe('gemini-pro');
    expect(response.data.fallback_occurred).toBe(true);
  });

  it('should handle validation errors', async () => {
    // Arrange - invalid request
    const invalidRequest = {
      script_content: '', // Empty content
      asset_types: []     // Empty asset types
    };

    const errorResponse = {
      error: 'script_content is required',
      error_code: 'VALIDATION_ERROR',
      timestamp: '2025-09-26T10:30:00Z'
    };

    mockedAxios.post.mockRejectedValueOnce({
      response: {
        status: 400,
        data: errorResponse
      }
    });

    // Act & Assert
    await expect(axios.post(`${baseURL}${endpoint}`, invalidRequest))
      .rejects.toMatchObject({
        response: {
          status: 400,
          data: errorResponse
        }
      });
  });

  it('should handle model unavailable scenarios', async () => {
    // Arrange - request when models unavailable
    const request: MediaGenerateRequest = {
      script_content: 'Test content',
      asset_types: ['image'],
      preferred_model: 'gemini-2.5-flash-image',
      allow_fallback: false
    };

    const unavailableResponse = {
      error: 'Preferred model unavailable and fallback disabled',
      error_code: 'MODEL_UNAVAILABLE',
      timestamp: '2025-09-26T10:30:00Z'
    };

    mockedAxios.post.mockRejectedValueOnce({
      response: {
        status: 503,
        data: unavailableResponse
      }
    });

    // Act & Assert
    await expect(axios.post(`${baseURL}${endpoint}`, request))
      .rejects.toMatchObject({
        response: {
          status: 503,
          data: unavailableResponse
        }
      });
  });

  it('should validate response structure completeness', async () => {
    // Arrange
    const request: MediaGenerateRequest = {
      script_content: 'Test script',
      asset_types: ['image', 'video_clip'],
      num_assets: 3,
      preferred_model: 'gemini-pro',
      allow_fallback: true
    };

    const completeResponse: MediaGenerateResponse = {
      job_id: '550e8400-e29b-41d4-a716-446655440003',
      status: 'pending',
      model_selected: 'gemini-pro',
      estimated_completion: '2025-09-26T10:40:00Z',
      fallback_occurred: false
    };

    mockedAxios.post.mockResolvedValueOnce({
      status: 202,
      data: completeResponse
    });

    // Act
    const response = await axios.post(`${baseURL}${endpoint}`, request);

    // Assert - validate all expected fields
    expect(response.data.job_id).toBeDefined();
    expect(response.data.status).toBeDefined();
    expect(response.data.model_selected).toBeDefined();

    // Validate UUID format for job_id
    const uuidRegex = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;
    expect(response.data.job_id).toMatch(uuidRegex);

    // Validate status enum
    const validStatuses = ['pending', 'generating', 'completed', 'failed'];
    expect(validStatuses).toContain(response.data.status);

    // Validate timestamp format if present
    if (response.data.estimated_completion) {
      expect(new Date(response.data.estimated_completion)).toBeInstanceOf(Date);
    }
  });
});

File: backend/tests/contract/test_media_generation_post.py
"""
Contract test for POST /api/media/generate endpoint
Task: T004 [P] - Contract test POST /api/media/generate
"""

import pytest
from fastapi.testclient import TestClient
from pydantic import ValidationError
from uuid import UUID
from datetime import datetime


class TestMediaGenerationPostContract:
    """Contract tests for POST /api/media/generate endpoint"""

    def test_valid_request_returns_202(self, client: TestClient):
        """Test valid media generation request returns 202 with correct schema"""
        request_data = {
            "script_content": "Speaker 1: Today we'll discuss AI advancements...",
            "asset_types": ["image"],
            "num_assets": 2,
            "preferred_model": "gemini-2.5-flash-image",
            "allow_fallback": True
        }

        response = client.post("/api/media/generate", json=request_data)

        # Should return 202 Accepted
        assert response.status_code == 202

        # Validate response schema
        data = response.json()
        assert "job_id" in data
        assert "status" in data
        assert "model_selected" in data
        assert "estimated_completion" in data

        # Validate specific fields
        UUID(data["job_id"])  # Should be valid UUID
        assert data["status"] in ["pending", "generating"]
        assert data["model_selected"] in ["gemini-2.5-flash-image", "gemini-pro"]
        datetime.fromisoformat(data["estimated_completion"].replace('Z', '+00:00'))

    def test_defaults_to_gemini_2_5_flash_image(self, client: TestClient):
        """Test request without preferred_model defaults to gemini-2.5-flash-image"""
        request_data = {
            "script_content": "Test content",
            "asset_types": ["image"]
        }

        response = client.post("/api/media/generate", json=request_data)

        if response.status_code == 202:
            data = response.json()
            assert data["model_selected"] == "gemini-2.5-flash-image"

    def test_invalid_asset_types_returns_400(self, client: TestClient):
        """Test invalid asset_types enum values return 400"""
        request_data = {
            "script_content": "Test content",
            "asset_types": ["invalid_type"],
            "num_assets": 1
        }

        response = client.post("/api/media/generate", json=request_data)
        assert response.status_code == 400

    def test_invalid_num_assets_range_returns_400(self, client: TestClient):
        """Test num_assets outside 1-10 range returns 400"""
        # Test below minimum
        request_data = {
            "script_content": "Test content",
            "asset_types": ["image"],
            "num_assets": 0
        }
        response = client.post("/api/media/generate", json=request_data)
        assert response.status_code == 400

        # Test above maximum
        request_data["num_assets"] = 11
        response = client.post("/api/media/generate", json=request_data)
        assert response.status_code == 400

    def test_missing_required_fields_returns_400(self, client: TestClient):
        """Test missing required fields return 400"""
        # Missing script_content
        response = client.post("/api/media/generate", json={"asset_types": ["image"]})
        assert response.status_code == 400

        # Missing asset_types
        response = client.post("/api/media/generate", json={"script_content": "test"})
        assert response.status_code == 400

    def test_invalid_preferred_model_returns_400(self, client: TestClient):
        """Test invalid preferred_model returns 400"""
        request_data = {
            "script_content": "Test content",
            "asset_types": ["image"],
            "preferred_model": "invalid-model"
        }

        response = client.post("/api/media/generate", json=request_data)
        assert response.status_code == 400

    def test_models_unavailable_returns_503(self, client: TestClient):
        """Test when all models unavailable returns 503"""
        # This test will pass when service properly handles model unavailability
        request_data = {
            "script_content": "Test content",
            "asset_types": ["image"],
            "allow_fallback": False
        }

        # This should eventually return 503 when models are unavailable
        response = client.post("/api/media/generate", json=request_data)
        # For now, just ensure it doesn't crash
        assert response.status_code in [202, 400, 503]


@pytest.fixture
def client():
    """Test client fixture - will fail until implementation exists"""
    from backend.main import app
    return TestClient(app)